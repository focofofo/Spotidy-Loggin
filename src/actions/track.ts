import { TrackPayload } from '../models'

export const SET_TRACKS = 'SET_TRACKS'
export const ALBUM_NOT_FOUND = 'ALBUM_NOT_FOUND'
export const ADD_FAVORITE_TRACK = 'ADD_FAVORITE_TRACK'
export const REMOVE_FAVORITE_TRACK = 'REMOVE_FAVORITE_TRACK'
export const CHECK_FAVORITES = 'CHECK_FAVORITES'
export const RESET_TRACKS = 'RESET_TRACKS'

/**
 * Set the album's tracks information into the storage
 * @param payload album's tracks information from the API
 */
export function doSetTracks(payload: TrackPayload) {
  return {
    type: SET_TRACKS,
    payload,
  }
}

/**
 * Indicates the error in case the album is not found
 */
export function doAlbumNotFound() {
  return {
    type: ALBUM_NOT_FOUND,
  }
}

/**
 * Update the is_favorite key of the specified track
 * @param payload index from the tracks storage
 */
export function doAddFavorite(payload: number) {
  return {
    type: ADD_FAVORITE_TRACK,
    payload,
  }
}

/**
 * Remove the is_favorite key of the specified track
 * @param payload index from the tracks storage
 */
export function doRemoveFavoriteTrack(payload: number) {
  return {
    type: REMOVE_FAVORITE_TRACK,
    payload,
  }
}

/**
 * Send to storage the liked songs information
 * @param payload array of booleans indicating if track is liked or not
 */
export function doCheckFavorites(payload: boolean[]) {
  return {
    type: CHECK_FAVORITES,
    payload,
  }
}

/**
 * Clear the tracks storage
 */
export function doResetTracks() {
  return {
    type: RESET_TRACKS,
  }
}
