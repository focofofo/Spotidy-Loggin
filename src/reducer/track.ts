import { AnyAction } from 'redux'
import {
  SET_TRACKS,
  ALBUM_NOT_FOUND,
  RESET_TRACKS,
  ADD_FAVORITE_TRACK,
  REMOVE_FAVORITE_TRACK,
  CHECK_FAVORITES,
} from '../actions/track'
import { Track, TrackPayload } from '../models'

const initialState: TrackPayload = {
  href: '',
  items: [],
  limit: 0,
  next: '',
  offset: 0,
  previous: '',
  total: 0,
  error: false,
}

export default function track(
  state: TrackPayload = initialState,
  action: AnyAction
) {
  switch (action.type) {
    case SET_TRACKS:
      return {
        ...state,
        ...action.payload,
      }
    case ALBUM_NOT_FOUND:
      return {
        ...state,
        error: true,
      }
    case ADD_FAVORITE_TRACK:
      state.items[action.payload].is_favorite = true
      return {
        ...state,
      }
    case REMOVE_FAVORITE_TRACK:
      state.items[action.payload].is_favorite = false
      return {
        ...state,
      }
    case CHECK_FAVORITES:
      state.items.map(
        (track: Track, index: number) =>
          (track.is_favorite = action.payload[index])
      )
      return {
        ...state,
      }
    case RESET_TRACKS:
      return {
        ...initialState,
      }
    default:
      return state
  }
}
