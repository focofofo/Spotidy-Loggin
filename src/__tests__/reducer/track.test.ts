import { AnyAction } from 'redux'
import track from '../../reducer/track'
import {
  SET_TRACKS,
  ALBUM_NOT_FOUND,
  ADD_FAVORITE_TRACK,
  REMOVE_FAVORITE_TRACK,
  CHECK_FAVORITES,
  RESET_TRACKS,
} from '../../actions/track'
import MockTrack from '../mock/Track.mock.json'

describe('track reducer', () => {
  it('should handle initial state', () => {
    expect(track(undefined, {} as AnyAction)).toMatchSnapshot()
  })

  it('should handle SET_TRACKS', () => {
    expect(
      track(undefined, {
        type: SET_TRACKS,
        payload: {
          href: 'test',
          items: [],
          limit: 10,
          next: 'test',
          offset: 10,
          previous: 'test',
          total: 10,
          error: false,
        },
      })
    ).toMatchSnapshot()
  })

  it('should handle ALBUM_NOT_FOUND', () => {
    expect(
      track(undefined, {
        type: ALBUM_NOT_FOUND,
      })
    ).toMatchSnapshot()
  })

  it('should handle ADD_FAVORITE_TRACK', () => {
    expect(
      track(
        {
          href: 'test',
          items: [MockTrack],
          limit: 10,
          next: 'test',
          offset: 10,
          previous: 'test',
          total: 10,
          error: false,
        },
        {
          type: ADD_FAVORITE_TRACK,
          payload: 0,
        }
      )
    ).toMatchSnapshot()
  })

  it('should handle REMOVE_FAVORITE_TRACK', () => {
    expect(
      track(
        {
          href: 'test',
          items: [MockTrack],
          limit: 10,
          next: 'test',
          offset: 10,
          previous: 'test',
          total: 10,
          error: false,
        },
        {
          type: REMOVE_FAVORITE_TRACK,
          payload: 0,
        }
      )
    ).toMatchSnapshot()
  })

  it('should handle CHECK_FAVORITES', () => {
    expect(
      track(
        {
          href: 'test',
          items: [MockTrack],
          limit: 10,
          next: 'test',
          offset: 10,
          previous: 'test',
          total: 10,
          error: false,
        },
        {
          type: CHECK_FAVORITES,
          payload: [false],
        }
      )
    ).toMatchSnapshot()
  })

  it('should handle RESET_TRACKS', () => {
    expect(track(undefined, { type: RESET_TRACKS })).toMatchSnapshot()
  })
})
