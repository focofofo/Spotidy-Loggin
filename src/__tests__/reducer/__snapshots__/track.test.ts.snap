// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`track reducer should handle ADD_FAVORITE_TRACK 1`] = `
Object {
  "error": false,
  "href": "test",
  "items": Array [
    Object {
      "artists": Array [],
      "disc_number": 0,
      "duration_ms": 0,
      "explicit": true,
      "external_urls": Object {},
      "href": "test",
      "id": "test",
      "is_favorite": true,
      "is_local": true,
      "is_playable": true,
      "name": "test",
      "track_number": 0,
      "type": "test",
      "uri": "test",
    },
  ],
  "limit": 10,
  "next": "test",
  "offset": 10,
  "previous": "test",
  "total": 10,
}
`;

exports[`track reducer should handle ALBUM_NOT_FOUND 1`] = `
Object {
  "error": true,
  "href": "",
  "items": Array [],
  "limit": 0,
  "next": "",
  "offset": 0,
  "previous": "",
  "total": 0,
}
`;

exports[`track reducer should handle CHECK_FAVORITES 1`] = `
Object {
  "error": false,
  "href": "test",
  "items": Array [
    Object {
      "artists": Array [],
      "disc_number": 0,
      "duration_ms": 0,
      "explicit": true,
      "external_urls": Object {},
      "href": "test",
      "id": "test",
      "is_favorite": false,
      "is_local": true,
      "is_playable": true,
      "name": "test",
      "track_number": 0,
      "type": "test",
      "uri": "test",
    },
  ],
  "limit": 10,
  "next": "test",
  "offset": 10,
  "previous": "test",
  "total": 10,
}
`;

exports[`track reducer should handle REMOVE_FAVORITE_TRACK 1`] = `
Object {
  "error": false,
  "href": "test",
  "items": Array [
    Object {
      "artists": Array [],
      "disc_number": 0,
      "duration_ms": 0,
      "explicit": true,
      "external_urls": Object {},
      "href": "test",
      "id": "test",
      "is_favorite": false,
      "is_local": true,
      "is_playable": true,
      "name": "test",
      "track_number": 0,
      "type": "test",
      "uri": "test",
    },
  ],
  "limit": 10,
  "next": "test",
  "offset": 10,
  "previous": "test",
  "total": 10,
}
`;

exports[`track reducer should handle RESET_TRACKS 1`] = `
Object {
  "error": false,
  "href": "",
  "items": Array [],
  "limit": 0,
  "next": "",
  "offset": 0,
  "previous": "",
  "total": 0,
}
`;

exports[`track reducer should handle SET_TRACKS 1`] = `
Object {
  "error": false,
  "href": "test",
  "items": Array [],
  "limit": 10,
  "next": "test",
  "offset": 10,
  "previous": "test",
  "total": 10,
}
`;

exports[`track reducer should handle initial state 1`] = `
Object {
  "error": false,
  "href": "",
  "items": Array [],
  "limit": 0,
  "next": "",
  "offset": 0,
  "previous": "",
  "total": 0,
}
`;
